### PR TITLE
Nonlinear Least Squares MathProgBase models

### PR DESCRIPTION
**Pull request by abelsiqueira**
Thanks, @mlubin and @odow, for the help.

The `MathProgNLSModel` receives a `JuMP` model with `NLexpression`s and a list of those expressions, creating two internal `JuMP` models, one with the expressions as constraints and the other as the equivalent minimization of `0.5*||F(x)||^2`. The residual, and its **sparse** Jacobian are obtained from the constraints of the first model. The second model provides the usual NLP functions, without having to use the residual - one exception is the Hessian-vector product function, due to this bug: https://github.com/JuliaOpt/JuMP.jl/issues/1204.

### CODE REVIEW

File: src/NLPModels.jl [hunk 1]
```diff
@@ -344,7 +344,7 @@ conscale(::AbstractNLPModel, ::AbstractVector) =
   throw(NotImplementedError("conscale"))
 
 if Pkg.installed("MathProgBase") != nothing
-  include("mpb_model.jl")
```
Comment (dpo) [inline src/NLPModels.jl]: What happened to `mpb_model`?

File: src/nlsmpb_model.jl [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+       obj, grad, grad!,
+       cons, cons!, jac_coord, jac, jprod, jprod!, jtprod, jtprod!,
+       hess_coord, hess, hprod, hprod!,
+       residual!, jac_residual, jprod_residual!, jtprod_residual!,
+       jac_op_residual, hess_residual, hprod_residual!, cons, cons!, jac_coord,
+       jac, jprod, jprod!, jtprod, jtprod!, hess, hess_coord, hprod, hprod!
+
+mutable struct MathProgNLSModel <: AbstractNLSModel
+  meta :: NLPModelMeta
+  nls_meta :: NLSMeta
```
Comment (dpo) [inline src/nlsmpb_model.jl]: Are we moving towards merging both kinds of models into one then?

File: src/nlsmpb_model.jl [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+function hprod!(nls :: MathProgNLSModel, x :: Array{Float64}, v :: Array{Float64},
+    hv :: Array{Float64};
+    obj_weight :: Float64=1.0, y :: Array{Float64}=zeros(nls.meta.ncon))
+  MathProgBase.eval_hesslag_prod(nls.cmodel.eval, hv, x, v, obj_weight, y)
+  return hv
+end
+=#
+
+function hprod!(nls :: MathProgNLSModel, x :: Array{Float64}, v :: Array{Float64},
+    hv :: Array{Float64};
```
Comment (dpo) [inline src/nlsmpb_model.jl]: So this version will be removed when 1204 is fixed?

**Comment by dpo** [comment]
This looks excellent, @abelsiqueira ! Many thanks!

File: src/NLPModels.jl (hunk 1 above)
Comment (abelsiqueira) [reply src/NLPModels.jl]: It's being `include`d by `nlsmpb_model.jl`. It uses many MPB things in common.

File: src/nlsmpb_model.jl (hunk 2 above)
Comment (abelsiqueira) [reply src/nlsmpb_model.jl]: Not yet, but this model will transition smoothly, it seems.

File: src/nlsmpb_model.jl (hunk 3 above)
Comment (abelsiqueira) [reply src/nlsmpb_model.jl]: Yes. Notice that this version increments `residual` and `j*prod_*`, while using `cmodel` doesn't require that.

File: src/NLPModels.jl (hunk 1 above)
Comment (dpo) [reply src/NLPModels.jl]: I missed that. Thanks!

File: src/nlsmpb_model.jl (hunk 2 above)
Comment (dpo) [reply src/nlsmpb_model.jl]: Ok

File: src/nlsmpb_model.jl (hunk 3 above)
Comment (dpo) [reply src/nlsmpb_model.jl]: 👍

**Pull request merged**